pydatastructs/miscellaneous_data_structures/queue.py

```diff
@@ -88,20 +89,40 @@ def __new__(cls, items=None, dtype=NoneType):
         obj.items, obj.front = items, -1
         if items.size == 0:
             obj.front = -1
+            obj.rear = -1
         else:
             obj.front = 0
+            obj.rear = items._num - 1
+        obj.deque = double_ended
         return obj
 
     @classmethod
     def methods(cls):
-        return ['__new__', 'append', 'popleft', 'rear',
-        'is_empty', '__len__', '__str__']
+        if cls.deque is True:
+            return ['__new__', 'append', 'appendleft', 'popleft', 'pop', 'is_empty', '__len__', '__str__']
+        else:
+            return ['__new__', 'append', 'popleft', 'is_empty', '__len__', '__str__']
 
     def append(self, x):
         if self.is_empty:
             self.front = 0
             self.items._dtype = type(x)
         self.items.append(x)
+        self.rear += 1
+
+    def appendleft(self, x):
+        if self.deque is False:
+            raise ImplementationError("Deque operations can't be done here")
+        temp = []
+        if self.is_empty:
+            self.front = 0
+            self.rear = -1
+            self.items._dtype = type(x)
+        temp.append(x)
+        for i in range(self.front, self.rear + 1):
+            temp.append(self.items._data[i])
```

Comment:
Copying the whole array to a new one?

Reply:
Yeah as the insertion at beginning involves movement of all the to its successive place and new element to be inserted in beginning, it will work in the same way as I did.